### ISSUE TITLE
Lint missing Copy impl/derive on fully public types.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#68497]
Any `pub enum`, and any `pub struct` which also has only `pub` fields, can be "manually copied", i.e. a new value of that type can be produced at any given time by copying out the data fields (of the active variant, in the `enum` case) and passing the values back to the constructor.

Therefore, we might want to lint cases in which such a type doesn't implement `Copy`, because the most likely reasons it doesn't are (assuming no `Drop` impl):
* it would be fine but was forgotten
* the fields weren't actually meant to be publicly accessible
  * this could be hinted at by phrasing the lint to express the fact that anyone can copy all the fields out and make a new value out of them

A good (and bad, at the same time) example is `std::ops::Range<T>`, which can be copied by simply doing `r.start..r.end` (or `Range { start: r.start, end: r.end }`, like any `struct`).

It's a bad example because the current (suboptimal) convention avoids `Copy` on types implementing `Iterator` (and I suppose the lint could take that into account), but if anyone else wrote a similar struct (and it wasn't an iterator), they would probably want it to be `Copy`.

cc @Manishearth @oli-obk

**Comment by dtolnay** [comment]
More than the reasons you listed, the biggest reason to not to have unnecessary impls on a pub type in my experience is compile time. Not specifically the Copy impls, but the requisite Clone impls are sometimes a big hit. For Syn, as measured by `cargo clean && time cargo build --no-default-features --features full` vs `--features full,clone-impls` with the only difference between those being some Clone impls, the build with Clone impls is slower by 10-15%. I think libc has a similar Cargo feature to gate extra traits but idk if it applies to Clone.

**Comment by eddyb** [comment]
I believe `#[derive(Copy, Clone)]` might be cheaper whenever the type isn't generic (but I could be wrong, or at least it might've changed since).

But alright, I'll add that reason, even though I'm really unhappy that such a tradeoff even exists, the alternative (if we don't "just" make rustc faster) is probably to skip typechecking and rely on MIR shims, which is not ideal either.

**Comment by petrochenkov** [comment]
We always had a built-in lint for this?
It's called `missing_copy_implementations`, it's just disabled by default.

**Comment by eddyb** [comment]
@petrochenkov Does it take field privacy into account, though?
I was aware of the lint (I should've made that clear though), but I was thinking of something aimed more at cases where "manual copying" is always an option, just an unergonomic one.

**Comment by petrochenkov** [comment]
No, it doesn't look at field privacy, only the struct's privacy.

**Comment by HeroicKatora** [comment]
Also consider `#[non_exhaustive]` on enums and structs. While the current fields and variants might all be copy, this lint should not be enabled by default here based on the description of the issue. It is not possible to copy them by matching and not implementing `Copy` could be motiviated by future compatibility, allowing the author to introduce non-`Copy` fields and variants.

**Comment by eddyb** [comment]
@HeroicKatora Thanks, I've added that to the description.
The intent is definitely to ignore cases like you described, where "manual copying by reconstructing the value with copies of all fields" isn't applicable.
